**HybPipe support files and scripts**  
In HybSeqSource folder there are all necessary files and scripts calling by main HybPipe bash scripts.  
Use of these scripts should be properly cited, e.g. as follows:  
  
**abacas.1.3.1.pl**  
http://abacas.sourceforge.net/Manual.html  
**AfterPhylo.pl**  
https://github.com/qiyunzhu/AfterPhylo  
**alignmentSummary.R**  
_part of HybPipe_  
**assembled_exons_to_fastas.py**  
https://github.com/listonlab/Hyb-Seq_protocol/blob/master/assembled_exons_to_fastas/assembled_exons_to_fastas.py  
**astral.4.10.2.jar**  
https://github.com/smirarab/ASTRAL  
**BaCoCa.v1.107.pl**  
https://www.zfmk.de/en/research/research-centres-and-groups/bacoca  
**catfasta2phyml.pl**  
https://github.com/nylander/catfasta2phyml  
**CompareToBootstrap.pl**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**CompareTree.pl**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**convert_unit_to_100_support.pl**  
https://github.com/smirarab/global/blob/master/src/perl/convert_unit_to_100_support.pl  
**create_beast_input.pl**  
http://www.cs.utexas.edu/users/phylo/datasets/bbca/create_beast_input.pl  
**cutfasta.pl**  
_part of HybPipe_  
**FASconCAT-G_v1.0.pl**  
https://www.zfmk.de/en/research/research-centres-and-groups/fasconcat  
**fastq2fasta.pl** 
http://brianknaus.com/software/srtoolbox/fastq2fasta.pl  
**histogram.r**  
_part of HybPipe_  
**LBscores.R**  
_part of HybPipe_  
**merge_FASTA.pl**  
**MOTree.pm**  
http://meta.microbesonline.org/fasttree/treecmp.html  
**mrp.jar**  
https://github.com/smirarab/mrpmatrix  
**NEBNext-PE.fa** 
Oligonucleotide sequences Â© 2006-2010 Illumina, Inc. All rights reserved. 
**PhiX.fsa**  
http://www.ncbi.nlm.nih.gov/nuccore/9626372  
**plotting_correlations.R**  
_part of HybPipe_  
**resolvepolytomies.R**  
_part of HybPipe_  
**tree_props.r**  
_part of HybPipe_  
**treepropsPlot.r**  
_part of HybPipe_  
**trimmomatic-0.33.jar**  
http://www.usadellab.org/cms/?page=trimmomatic  
